language/vm/vm_runtime/src/txn_executor.rs

```diff
@@ -254,39 +256,56 @@ where
                     if callee_function_ref.is_native() {
                         let module_name: &str = callee_function_ref.module().name();
                         let function_name: &str = callee_function_ref.name();
-                        let native_function =
-                            match dispatch_native_function(module_name, function_name) {
-                                None => return Err(VMInvariantViolation::LinkerError),
-                                Some(native_function) => native_function,
-                            };
-                        let mut arguments = VecDeque::new();
-                        for _ in 0..native_function.num_args() {
-                            arguments.push_front(self.execution_stack.pop()?);
-                        }
-                        let (cost, return_values) = match (native_function.dispatch)(arguments) {
-                            NativeReturnStatus::InvalidArguments => {
-                                // TODO: better error
-                                return Err(VMInvariantViolation::LinkerError);
+                        if module_name == EVENT_NAME && function_name == EMIT_EVENT_NAME {
```

Comment:
You'll need to check the address here too. 

Nonetheless, this makes me deeply uncomfortable. The purpose of the `native_functions` module is to encapsulate the natives in a safe manner. 

I see two possible ways forward
1) Still use `dispatch_native_function` but then do not call `native_function.dispatch`. At least in this way it's still using the same encapsulation 
2) Add another field to `NativeReturnStatus` for emitting event data. 

(2) seems potentially dangerous as we really don't want other natives emitting events
(1) is kind of gross but feels safer at least that any checks around lookup/linking of natives will  use the same mechanism

Reply:
I prefer option 1 better. Will do that!